mantidproject/mantid
# Add CMake export header feature to more targets

**Description of work.**

Use the cmake GenerateExportHeader feature on some further code (DataHandling, DataObjects and the joint EventParallelLoader\Parallel target). This means that cmake generates the `MANTID_<TargetName>_DLL` macro (eg `MANTID_DATAHANDLING_DLL`) instead of it being done inside Mantid code. This macro gets expanded to `__declspec(dllexport)` or `__declspec(dllexport)` on Windows depending on the context. And on linux it gets expanded to `__attribute__(visibility("default"))` or blank depending on the context.

While doing this I have also ensured the `MANTID_<TargetName>_DLL` macro is used on all class definitions apart from template classes. In some places `DLLExport` was used instead which meant the wrong macro was present if a header file was imported into another library.

This is being done in chunks across several maintenance periods so this just applies it to some more targets\libraries rather than doing the whole code base.

**To test:**

No functional change so just code review and check the Jenkins jobs all pass

Although lots of files changed, the only interesting edits are to the CMakeLists.txt files. The remaining edits are mainly changes to use the `MANTID_<TargetName>_DLL` macro instead of the `DLLExport` macro

Fixes #27061.

*This does not require release notes* because **no functional change**



---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
